Lift -triton-raise-block-pointer pass from intel-xpu-backend-for-triton

rolfmorel opened the pull request:
Code was in turn taken from triton-shared (though does not use the tts dialect).

adam-smnk commented on third_party/cpu/include/TritonRaiseBlockPointer/Passes.td:
```diff
@@ -0,0 +1,30 @@
+include "mlir/Pass/PassBase.td"
+
+def TritonRaiseBlockPointer
+    : Pass<"triton-raise-block-pointer", "mlir::ModuleOp"> {
```
nit: `triton-cpu-`

rolfmorel replied:
Done.

adam-smnk commented:
Atm it doesn't really compose with triton-cpu rewrites in the way I'd expect.
When input kernel has hand written block ptrs, `triton-cpu-convert-memory-ops` manages to translate `tt.load` into 2D `vector.transfer_read`. After raising from scalar loads, the loads get split into unrolled vector instructions so the source memref is lost at this point.

adam-smnk commented:
Actually `canonicalize` undoes the raising back into `tt.splat` and `tt.addptr` so imo the raising is pretty poor.

adam-smnk commented:
Ohhh now I see looking closer at IR. Created `tt.make_tensor_ptr` is unused. No wonder it gets removed.

adam-smnk commented:
> Ohhh now I see looking closer at IR. Created `tt.make_tensor_ptr` is unused. No wonder it gets removed.

After an offline discussion, we found the issue. Raising fails to kick in for a kernel with fully `constexpr` arguments. It does seem to work on unmodified `03-gemm` example.

rolfmorel commented:
Rebased, added `-cpu` to pass name (and added MLIR arg names to kernel args in `03-matrix-multiplication-cpu.py`) and made the raising parameterized.

The raising indeed seems to work on unmodified `03-gemm` but appears a bit fragile.

rolfmorel commented on python/tutorials/03-matrix-multiplication-cpu.py:
```diff
@@ -328,7 +335,7 @@ def matmul(a: torch.Tensor, b: torch.Tensor, c: torch.Tensor):
 torch.manual_seed(0)
 
-triton.runtime.driver.set_active_to_cpu(experimental=True)
+triton.runtime.driver.set_active_to_cpu(experimental=False)
```
With `experimental=True` the V2 backend is used for the correctness check. We modify the V1 pipeline. Note that the correctness check only runs once, for one backend/pipeline. 

My mistake for not highlighting the change - I forgot. Nevertheless, I think we should change this (a separate PR would be more appropriate, but I don't think we need to spend time on all that).

adam-smnk replied:
Ohhh that totally makes sense and explains why it's been so slow recently.
Let's absolutely leave it changed to V1.

adam-smnk replied:
On that note, revalidated both XSMM paths. Everything's still good.

rolfmorel commented:
Indeed, no changes to the raising logic. Just moving some code over and slotting in some utility functions.
